src/wp-admin/includes/image.php

```diff
@@ -301,47 +318,91 @@ function wp_create_image_subsizes( $file, $attachment_id ) {
 			if ( ! is_wp_error( $resized ) ) {
 				// Append "-scaled" to the image file name. It will look like "my_image-scaled.jpg".
 				// This doesn't affect the sub-sizes names as they are generated from the original image (for best quality).
-				$saved = $editor->save( $editor->generate_filename( 'scaled' ) );
+				$saved = $editor->save( $editor->generate_filename( 'scaled' ), $primary_mime_type );
 
 				if ( ! is_wp_error( $saved ) ) {
 					$image_meta = _wp_image_meta_replace_original( $saved, $file, $image_meta, $attachment_id );
 
+					$image_meta['sources'][ $primary_mime_type ] = _wp_get_sources_from_meta( $saved );
+
 					// If the image was rotated update the stored EXIF data.
 					if ( true === $rotated && ! empty( $image_meta['image_meta']['orientation'] ) ) {
 						$image_meta['image_meta']['orientation'] = 1;
 					}
+					wp_update_attachment_metadata( $attachment_id, $image_meta );
 				} else {
 					// TODO: Log errors.
 				}
+
+				// Save additional mime types.
+				foreach ( $additional_mime_types as $output_mime_type ) {
+					$extension = wp_get_default_extension_for_mime_type( $output_mime_type );
+					// @todo Correct issue where filename winds up with duplicate "-scaled" suffix added unless the editor is reset.
+					// adding extension to name for now.
+					$saved = $editor->save( $editor->generate_filename( $extension ), $output_mime_type );
```

Comment:
That makes sense and will be a nice improvement, simplifying the approach. This lines up nicely with what you were suggesting before as an alternate approach. I'll give it a try!

Reply:
@felixarntz implemented this approach in [`99cd2e4` (#2393)](https://github.com/WordPress/wordpress-develop/pull/2393/commits/99cd2e4b584a8a2d54022e47aefd8efeb1cb09e9) - really like this, makes change to image class automatically compatible since any custom classes build from the base class. No changes are required in core's Imagick and GD classes now :) 

Unfortunately, it isn't working calling save directly works, but make_subsize fails to output the correct mime type :( 

* The call to `make_subsize` calls `_save` with only an image, (not passing filename or mime type). 
* `_save` in turn calls `get_output_format` to determine the output format. 
* This function in the base class falls back to using the (main) file extension, therefore this function returns the filename as ".jpg" although the mime type and ext are set correctly.
* 
What do you think, should we consider setting

Reply:
(reverting changes for now since it didn't work as expected, please test from [th](https://github.com/WordPress/wordpress-develop/commit/00dbbf9626dafe581940697bf6d75046ca569d10)

Reply:
@adamsilverstein 
Ah, good point on the extension. Maybe we could turn that around though - we can adjust the base class method since it's in the base class.

Basically:
* After setting `$file_ext` and `$file_mime` like today, call the `get_extension` method based on `$file_mime`.
* If that extension is different from `$file_ext`, use that one instead.
* Potentially we could run this extra logic only if the MIME type has been manually adjusted via `set_mime_type`. From that perspective, it might be good to also introduce another method `reset_mime_type` that just sets it back to the MIME type based on the `$file`. And it could set another new private variable, just a flag whether the MIME type was manually adjusted or not.

Reply:
Ok, that makes sense, i'll give that a try. making the change in the base class is certainly the best bet for compatibility and I can see how we need the safety check since the mime type can also be changed when saving which is already filterable.

Reply:
Got this working in [`98ab980` (#2393)](https://github.com/WordPress/wordpress-develop/pull/2393/commits/98ab98017e0814b6259e367ae3dec02513957a5b)